Sources/Networking/Documentation.docc/Documentation.md

```diff
@@ -0,0 +1,297 @@
+
+3. Invalidate download session in case DownloadAPIManager is not used as singleton to prevent memory leaks.
+```swift
+func invalidateSession(shouldFinishTasks: Bool = false)
+```
+DownloadAPIManager is not deallocated from memory since URLSession is holding a reference to it. If you wish to use new instances of the DownloadAPIManager, don't forget to invalidate the session if it is not needed anymore.
+
+## Retry ability
+Both APIManager and DownloadAPIManager allow for configurable retry mechanism.
+
+```swift
+let retryConfiguration = RetryConfiguration(retries: 2, delay: .constant(1)) { error in 
+    // custom logic here that determines whether the request should be retried
+    // e.g you can only retry with 5xx http error codes
+}
+```
+
+## Interceptors
+Interceptors are useful pieces of code that modify request/response in the network request pipeline.
+![Interceptors diagram](interceptors-diagram.png)
+
+There are three types you can leverage:<br>
+
+``RequestAdapting``
+
+Adapters are request transformable components that perform operations on the URLRequest before it is dispatched. They are used to further customise HTTP requests before they are carried out by editing the URLRequest (e.g updating headers).
+
+``ResponseProcessing``
+
+Processors are modifying the URLResponse received after a successful network request.
+
+``RequestInterceptor``
+
+Interceptors handle both adapting and processing.
+
+By conforming to these protocols, you can create your own adaptors/processors/interceptors. In the following part, interceptors provided by Networking are introduced.
+
+## Request Interceptors
+
+### Logging
+Networking provides a default ``LoggingInterceptor`` which internally uses `os_log` to pretty print requests/responses. You can utilise it to get logging console output either for requests, responses or both.
+
+```swift
+APIManager(
+    //
+    requestAdapters: [LoggingInterceptor.shared],
+    responseProcessors: [LoggingInterceptor.shared],
+    errorProcessors: [LoggingInterceptor.shared]
+    //
+)
+```
+
+### Authorization
+Networking provides a default authorization handling for OAuth scenarios. Use the default ``AuthorizationTokenInterceptor`` with the APIManager to obtain the behaviour of JWT Bearer authorization header injection and access token expiration refresh flow.
+
+Start by implementing an authorization manager by conforming to ``AuthorizationManaging``. This manager requires you to provide storage defined by ``AuthorizationStorageManaging`` (where OAuth credentials will be stored) and a refresh method that will perform the refresh token network call to obtain a new OAuth pair. Optionally, you can provide custom implementations for ``AuthorizationManaging/authorizeRequest(_:)-6azlk`` (by default, this method sets the authorization header) or access token getter (by default, this method returns the access token saved in provided storage).
```

Comment:
I would add a note about the fact that if access token is expired or api returns 401 networking library automatically tries to refresh access token